proxyclient/m1n1/shell.py

```diff
@@ -82,11 +110,29 @@ def display(val):
 
                 member = getattr(obj_class, attr)
                 if callable(member) and not isinstance(member, property):
-                    locals[attr] = getattr(obj, attr)
+                    cmd = getattr(obj, attr)
+                    locals[attr] = cmd
 
         for attr in dir(sysreg):
             locals[attr] = getattr(sysreg, attr)
 
+        for obj_name in locals.keys():
+            obj = locals.get(obj_name)
+            if obj is None:
+                continue
+            if callable(obj) and not isinstance(obj, property):
+                desc = locals[obj_name].__doc__
+                if not desc:
```

Comment:
What's this `re` stuff doing, exactly? If you want to get the class name and such, there are easier ways to do that without resorting to string parsing :)

Reply:
Please point me in the right direction. I thought there would be cleaner methods but I couldn't see any so a pointer would be very helpful.

Reply:
@amworsley Maybe [lib/inspect](https://docs.python.org/3/library/inspect.html) can be helpful? I see it has a way to check if something is a function — `inspect.isfunction()` — and a way to access it's qualified name — `obj. __qualname__ `. Hope that's the right direction!

Reply:
That looks fantastic, exactly the sort of thing I need.

Thanks very much Ludwig, I will have another go later today.

I had been working on another branch adding in some commands to dump the page tables and physical memory this item has blocked me from submitting
a new PR for it. So thanks again

Andrew

Reply:
Ok - switched to using the inspect class to get the Signature / command names and split the commands into categories so that it is much less cluttered.

See below for an example output:

```
..........................................................................................
Copying SEPFW (0x750000 bytes)...
Copying TrustCache (0x60000 bytes)...
Adjusting addresses in ADT...
Setting up bootargs at 0x81c0fc000...
Entering hypervisor shell. Type `start` to start the guest.
>>> help
Note: To display a category's commands quote the name e.g. help('HV')
List of top level commands:
TraceMode  : TraceMode(value, names=None, *, module=None, qualname=None, type=None, start=1) - 
HV         : HV category - 50 subcommands
map_hw     : map_hw(ipa, pa, size) - map IPA (Intermediate Physical Address) to actual PA
h          : h(number, /) - Return the hexadecimal representation of an integer.
UartInterface:
             UartInterface category - 17 subcommands
M1N1Proxy  : M1N1Proxy category - 108 subcommands
clear16    : clear16(addr, data) - Clear bits in 16 bit memory at address addr that are set
clear32    : clear32(addr, data) - Clear bits in 32 bit memory at address addr that are set
clear64    : clear64(addr, data) - Clear bits in 64 bit memory at address addr that are set
clear8     : clear8(addr, data) - Clear bits in 8 bit memory at addr that are set in data
hv_translate:
             hv_translate(addr, s1=False, w=False) - Translate virtual address
mask32     : mask32(addr, clear, set) - Clear bits in 32 bit memory at address addr that are
mask64     : mask64(addr, clear, set) - Clear bits in 64 bit memory at address addr that are
mask8      : mask8(addr, clear, set) - Clear bits in 1 byte memory at addr that are set
mset64ask16:
             mset64ask16(addr, clear, set) - Clear select bits in 16 bit memory addr that are set
read16     : read16(addr) - return 2 byte value from given address
read32     : read32(addr) - return 4 byte value given address
read64     : read64(addr) - return 8 byte value from given address
read8      : read8(addr) - return 1 byte value from given address
set16      : set16(addr, data) - Or 16 bit value of data into memory at addr and return result
set32      : set32(addr, data) - Or 32 bit value of data into memory at addr and return result
set64      : set64(addr, data) - Or 64 bit value of data into memory at addr and return result
set8       : set8(addr, data) - Or byte value of data into memory at addr and return result
write16    : write16(addr, data) - write 2 byte value to given address
write32    : write32(addr, data) - write 4 byte value to given address
write64    : write64(addr, data) - write 8 byte value to given address
write8     : write8(addr, data) - write 1 byte value to given address
ProxyUtils : ProxyUtils category - 16 subcommands
disassemble_at:
             disassemble_at(start, size, pc=None) - disassemble len bytes of memory from start
dump_mem   : dump_mem(ba=34359738368, n=48) - dump_mem(ba=0x8_0000_0000, len=48) - hex dump n bytes memory from ba
dump_pt    : dump_pt(arg=None, debug=None) -  dump_pt(debug=level) - dump translation tables
mrs        : mrs(reg, *, silent=False, call=None) - read system register reg
msr        : msr(reg, val, *, silent=False, call=None) - Write val to system register reg
read       : read(addr, width) - do a width read from addr and return it
write      : write(addr, data, width) - do a width write of data to addr
DABORT_DFSC:
             DABORT_DFSC(value, names=None, *, module=None, qualname=None, type=None, start=1) - An enumeration.
ESR_EC     : ESR_EC(value, names=None, *, module=None, qualname=None, type=None, start=1) - An enumeration.
Enum       : Enum(value, names=None, *, module=None, qualname=None, type=None, start=1) - 
IntEnum    : IntEnum(value, names=None, *, module=None, qualname=None, type=None, start=1) - Enum where members are also (and must be) ints
MSR_DIR    : MSR_DIR(value, names=None, *, module=None, qualname=None, type=None, start=1) - An enumeration.
SPSR_M     : SPSR_M(value, names=None, *, module=None, qualname=None, type=None, start=1) - An enumeration.
```